File: src/cli/commands/lint.ts
import { readFileSync } from "fs";
import { resolve } from "path";
import type { Command } from "commander";
import { tokenize } from "../../parser/index.js";
import { parse } from "../../parser/index.js";
import { buildGraph } from "../../parser/index.js";
import { inferShapes } from "../../shape/infer.js";
import { registry } from "../../blocks/registry.js";
import { lint, type LintDiagnostic } from "../../lint/index.js";
import { loadPlugins } from "../../plugins/index.js";
import { loadConfig } from "../../config/loader.js";
// Import side-effect: registers all built-in blocks
import "../../blocks/index.js";

function formatDiagnostic(d: LintDiagnostic): string {
  const { severity, message, loc, rule } = d;
  const locStr = `${loc.line}:${loc.col}`;
  return `${severity.toUpperCase()} [${rule}] at ${locStr}: ${message}`;
}

export function registerLintCommand(program: Command): void {
  program
    .command("lint <file>")
    .description("Lint a .mlmd file for errors and warnings")
    .option("--json", "output diagnostics as JSON")
    .action(async (file: string, options: { json?: boolean }) => {
      // Load plugins from config if present
      const config = loadConfig();
      if (config?.plugins) {
        await loadPlugins(config.plugins);
      }

      let source: string;
      try {
        source = readFileSync(file, "utf-8");
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        console.error(`Error reading file: ${msg}`);
        process.exit(1);
      }

      const tokens = tokenize(source);
      const parseResult = parse(tokens);

      if (parseResult.errors.length > 0) {
        if (options.json) {
          const jsonDiags = parseResult.errors.map((e) => ({
            severity: "error",
            message: e.message,
            loc: e.loc,
            rule: "parse-error",
          }));
          console.log(JSON.stringify(jsonDiags, null, 2));
        } else {
          for (const e of parseResult.errors) {
            const locStr = `${e.loc.line}:${e.loc.col}`;
            console.log(`ERROR [parse-error] at ${locStr}: ${e.message}`);
          }
        }
        process.exit(1);
      }

      const graph = buildGraph(parseResult.nodes);
      const shapeResult = inferShapes(graph, registry);

      // Collect shape errors too
      const allDiags: LintDiagnostic[] = shapeResult.errors.map((e) => ({
        severity: "error",
        message: e.message,
        loc: { line: 0, col: 0, offset: 0 },
        rule: "shape-mismatch",
      }));

      const lintDiags = lint(graph, registry);
      allDiags.push(...lintDiags);

      if (options.json) {
        console.log(JSON.stringify(allDiags, null, 2));
      } else {
        if (allDiags.length === 0) {
          console.log("No issues found.");
        } else {
          for (const d of allDiags) {
            console.log(formatDiagnostic(d));
          }
        }
      }

      const hasErrors = allDiags.some((d) => d.severity === "error");
      process.exit(hasErrors ? 1 : 0);
    });
}

File: mlmd/src/commands/lint.rs
use clap::Args;

use mlmd_core::block::registry::build_block_registry;
use mlmd_core::lint::{lint, LintDiagnostic, Severity};
use mlmd_core::parser::graph_builder::build_graph;
use mlmd_core::parser::parser::parse;
use mlmd_core::parser::tokenizer::tokenize;

#[derive(Args)]
pub struct LintArgs {
    /// Input .mlmd file(s)
    pub files: Vec<String>,
    /// Output diagnostics as JSON
    #[arg(long)]
    pub json: bool,
}

pub fn run(args: LintArgs) -> anyhow::Result<()> {
    let registry = build_block_registry();

    let mut all_diags: Vec<(String, Vec<LintDiagnostic>)> = Vec::new();
    let mut exit_code: i32 = 0;

    for file in &args.files {
        let source = std::fs::read_to_string(file)
            .map_err(|e| anyhow::anyhow!("Failed to read {}: {}", file, e))?;

        let tokens = tokenize(&source);
        let parse_result = parse(&tokens);

        // Report parse errors
        for err in &parse_result.errors {
            eprintln!(
                "ERROR [parse] at {}:{}: {}",
                err.loc.line, err.loc.col, err.message
            );
            exit_code = 1;
        }

        if parse_result.nodes.is_empty() {
            if !parse_result.errors.is_empty() {
                // Nodes are empty due to errors, skip further checks
                continue;
            }
            // Empty file — no diagnostics
            continue;
        }

        let graph = build_graph(&parse_result.nodes);
        let diags = lint(&graph, &registry);

        if diags.is_empty() && parse_result.errors.is_empty() {
            continue;
        }

        all_diags.push((file.clone(), diags));
    }

    if args.json {
        let json_output: Vec<serde_json::Value> = all_diags
            .iter()
            .flat_map(|(file, diags)| {
                diags.iter().map(move |d| {
                    serde_json::json!({
                        "file": file,
                        "severity": match d.severity {
                            Severity::Error => "error",
                            Severity::Warning => "warning",
                        },
                        "rule": d.rule,
                        "line": d.loc.line,
                        "col": d.loc.col,
                        "message": d.message,
                    })
                })
            })
            .collect();

        println!("{}", serde_json::to_string_pretty(&json_output)?);
    } else {
        for (_file, diags) in &all_diags {
            for d in diags {
                let severity_str = match d.severity {
                    Severity::Error => "ERROR",
                    Severity::Warning => "WARNING",
                };
                println!(
                    "{} [{}] at {}:{}: {}",
                    severity_str, d.rule, d.loc.line, d.loc.col, d.message
                );
                if d.severity == Severity::Error {
                    exit_code = 1;
                }
            }
        }
    }

    if exit_code != 0 {
        std::process::exit(exit_code);
    }

    Ok(())
}
